core/templates/domain/statistics/LearnerActionObjectFactory.ts

```diff
@@ -20,96 +20,168 @@
+}
+
+export interface ILearnerActionAnswerSubmitBackendDict {
+  // NOTE TO DEVELOPERS: The following two definitions are NOT typos! These are
+  // discriminated unions, a feature of TypeScript. Code will fail to compile if
+  // the compiler can not deduce that action_type and schema_version are set to
+  // the following values. For details see:
+  // https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions
+  'action_type': 'AnswerSubmit';
+  'schema_version': 1;
+  'action_customization_args': IAnswerSubmitCustomizationArgs;
+}
+
+export interface ILearnerActionExplorationQuitBackendDict {
+  // NOTE TO DEVELOPERS: The following two definitions are NOT typos! These are
+  // discriminated unions, a feature of TypeScript. Code will fail to compile if
+  // the compiler can not deduce that action_type and schema_version are set to
+  // the following values. For details see:
+  // https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions
+  'action_type': 'ExplorationQuit';
+  'schema_version': 1;
+  'action_customization_args': IExplorationQuitCustomizationArgs;
+}
+
+export type ILearnerActionBackendDict = (
+  ILearnerActionExplorationStartBackendDict |
+  ILearnerActionAnswerSubmitBackendDict |
+  ILearnerActionExplorationQuitBackendDict);
+
+export class LearnerActionExplorationStart {
+  actionType: 'ExplorationStart' = 'ExplorationStart';
+  schemaVersion: 1 = 1;
+  actionCustomizationArgs: IExplorationStartCustomizationArgs;
+
+  constructor(actionCustomizationArgs: IExplorationStartCustomizationArgs) {
+    this.actionCustomizationArgs = actionCustomizationArgs;
+  }
+
+  toBackendDict(): ILearnerActionExplorationStartBackendDict {
+    return {
+      action_type: this.actionType,
+      schema_version: this.schemaVersion,
+      action_customization_args: this.actionCustomizationArgs,
     }
+  }
+}
+
+export class LearnerActionAnswerSubmit {
+  actionType: 'AnswerSubmit' = 'AnswerSubmit';
+  schemaVersion: 1 = 1;
+  actionCustomizationArgs: IAnswerSubmitCustomizationArgs;
 
-    /** @type {string} */
-    this.actionType = actionType;
-    /** @type {Object.<string, *>} */
+  constructor(actionCustomizationArgs: IAnswerSubmitCustomizationArgs) {
     this.actionCustomizationArgs = actionCustomizationArgs;
-    /** @type {number} */
-    this.schemaVersion = schemaVersion;
   }
 
-  // TODO(#7176): Replace 'any' with the exact type. This has been kept as
-  // 'any' because the return type is a dict with underscore_cased
-  // keys which give tslint errors against underscore_casing in favor of
-  // camelCasing.
-  toBackendDict(): any {
+  toBackendDict(): ILearnerActionAnswerSubmitBackendDict {
     return {
       action_type: this.actionType,
+      schema_version: this.schemaVersion,
       action_customization_args: this.actionCustomizationArgs,
+    };
+  }
+}
+
+export class LearnerActionExplorationQuit {
+  actionType: 'ExplorationQuit' = 'ExplorationQuit';
```

Comment:
Mm ... what is going on here? I don't really understand how to interpret this.

/cc @nishantwrp @vojtechjelinek -- is this the recommended approach?

Reply:
This was confusing to me too, the "type" of this variable is the string value 'ExplorationQuit', and the "value" is the string value 'ExplorationQuit'. It needs both, otherwise the variable has an undefined value.

Reply:
Hm. What do other domain objects across the codebase do? Can we follow the existing style?

Reply:
The rest of the codebase doesn't have the strong type safety this provides, seems worth a bit of extra duplication..?

Reply:
I think the question is what convention we want to apply across the entire codebase. We need the whole codebase to be consistent, otherwise it will confuse developers.

So maybe this is the convention we want to apply throughout, but I think you should probably talk with @nishantwrp and @vojtechjelinek and reach a common decision. If the desire is to do something like this then we need to figure out a plan for doing it across the whole codebase too, before starting to implement it in a local area.

Reply:
OK, I can concede for this PR since I'd still like the PR to go in with or without this convention. I've filed #9502 where we can continue discussions, and will remove the changes from this PR.

Reply:
Obsoleted, no longer using discriminated unions.